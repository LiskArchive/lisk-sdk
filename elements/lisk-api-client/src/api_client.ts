/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as os from 'os';
import { HashMap, InitOptions } from './api_types';
import * as constants from './constants';
import { AccountsResource } from './resources/accounts';
import { BlocksResource } from './resources/blocks';
import { DappsResource } from './resources/dapps';
import { DelegatesResource } from './resources/delegates';
import { NodeResource } from './resources/node';
import { PeersResource } from './resources/peers';
import { SignaturesResource } from './resources/signatures';
import { TransactionsResource } from './resources/transactions';
import { VotersResource } from './resources/voters';
import { VotesResource } from './resources/votes';

const defaultOptions = {
	bannedNodes: [],
	randomizeNodes: true,
};

const commonHeaders: HashMap = {
	Accept: 'application/json',
	'Content-Type': 'application/json',
};

const getClientHeaders = (clientOptions: ClientOptions): HashMap => {
	const { name = '????', version = '????', engine = '????' } = clientOptions;

	const liskElementsInformation =
		'LiskElements/1.0 (+https://github.com/LiskHQ/lisk-elements)';
	const locale: string | undefined =
		process.env.LC_ALL ||
		process.env.LC_MESSAGES ||
		process.env.LANG ||
		process.env.LANGUAGE;
	const systemInformation = `${os.platform()} ${os.release()}; ${os.arch()}${
		locale ? `; ${locale}` : ''
	}`;

	return {
		'User-Agent': `${name}/${version} (${engine}) ${liskElementsInformation} ${systemInformation}`,
	};
};

export interface ClientOptions {
	readonly engine?: string;
	readonly name?: string;
	readonly version?: string;
}

export class APIClient {
	public static get constants(): typeof constants {
		return constants;
	}

	public static createMainnetAPIClient(options?: InitOptions): APIClient {
		return new APIClient(constants.MAINNET_NODES, {
			nethash: constants.MAINNET_NETHASH,
			...options,
		});
	}

	public static createTestnetAPIClient(options?: InitOptions): APIClient {
		return new APIClient(constants.TESTNET_NODES, {
			nethash: constants.TESTNET_NETHASH,
			...options,
		});
	}

	public accounts: AccountsResource;
	public bannedNodes!: ReadonlyArray<string>;
	public blocks: BlocksResource;
	public currentNode!: string;
	public dapps: DappsResource;
	public delegates: DelegatesResource;
	public headers!: HashMap;
	public node: NodeResource;
	public nodes!: ReadonlyArray<string>;
	public peers: PeersResource;
	public randomizeNodes!: boolean;
	public signatures: SignaturesResource;
	public transactions: TransactionsResource;
	public voters: VotersResource;
	public votes: VotesResource;

	public constructor(
		nodes: ReadonlyArray<string>,
		providedOptions: InitOptions = {},
	) {
		this.initialize(nodes, providedOptions);
		this.accounts = new AccountsResource(this);
		this.blocks = new BlocksResource(this);
		this.dapps = new DappsResource(this);
		this.delegates = new DelegatesResource(this);
		this.node = new NodeResource(this);
		this.peers = new PeersResource(this);
		this.signatures = new SignaturesResource(this);
		this.transactions = new TransactionsResource(this);
		this.voters = new VotersResource(this);
		this.votes = new VotesResource(this);
	}

	public banActiveNode(): boolean {
		return this.banNode(this.currentNode);
	}

	public banActiveNodeAndSelect(): boolean {
		const banned = this.banActiveNode();
		if (banned) {
			this.currentNode = this.getNewNode();
		}

		return banned;
	}

	public banNode(node: string): boolean {
		if (!this.isBanned(node)) {
			this.bannedNodes = [...this.bannedNodes, node];

			return true;
		}

		return false;
	}

	public getNewNode(): string {
		const nodes = this.nodes.filter(
			(node: string): boolean => !this.isBanned(node),
		);

		if (nodes.length === 0) {
			throw new Error('Cannot get new node: all nodes have been banned.');
		}

		const randomIndex = Math.floor(Math.random() * nodes.length);

		return nodes[randomIndex];
	}

	public hasAvailableNodes(): boolean {
		return this.nodes.some((node: string): boolean => !this.isBanned(node));
	}

	public initialize(
		nodes: ReadonlyArray<string>,
		providedOptions: InitOptions = {},
	): void {
		if (!Array.isArray(nodes) || nodes.length <= 0) {
			throw new Error('APIClient requires nodes for initialization.');
		}

		if (typeof providedOptions !== 'object' || Array.isArray(providedOptions)) {
			throw new Error(
				'APIClient takes an optional object as the second parameter.',
			);
		}

		const options: InitOptions = { ...defaultOptions, ...providedOptions };

		this.headers = {
			...commonHeaders,
			...(options.nethash ? { nethash: options.nethash } : {}),
			...(options.client ? getClientHeaders(options.client) : {}),
		};

		this.nodes = nodes;
		this.bannedNodes = [...(options.bannedNodes || [])];
		this.currentNode = options.node || this.getNewNode();
		this.randomizeNodes = options.randomizeNodes !== false;
	}

	public isBanned(node: string): boolean {
		return this.bannedNodes.includes(node);
	}
}
