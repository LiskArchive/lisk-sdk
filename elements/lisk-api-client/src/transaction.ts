/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import {
	signTransaction,
	signMultiSignatureTransaction,
	computeMinFee,
} from '@liskhq/lisk-transactions';
import {
	getAddressAndPublicKeyFromPassphrase,
	getAddressFromPublicKey,
} from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import {
	decodeTransaction,
	encodeTransaction,
	getTransactionAssetSchema,
	decodeAccount,
} from './codec';
import { Channel, RegisteredSchemas, NodeInfo } from './types';

interface MultiSignatureKeys {
	readonly mandatoryKeys: Buffer[];
	readonly optionalKeys: Buffer[];
	readonly numberOfSignatures: number;
}

export class Transaction {
	private readonly _channel: Channel;
	private readonly _schema: RegisteredSchemas;
	private readonly _nodeInfo: NodeInfo;

	public constructor(channel: Channel, registeredSchema: RegisteredSchemas, nodeInfo: NodeInfo) {
		this._channel = channel;
		this._schema = registeredSchema;
		this._nodeInfo = nodeInfo;
	}

	public async create(
		input: {
			moduleID?: number; // id takes priority
			moduleName?: string;
			assetID?: number; // id takes priority
			assetName?: string;
			fee: bigint;
			nonce?: bigint;
			senderPublicKey?: Buffer;
			asset: Record<string, unknown>;
			signatures?: Buffer[];
		},
		passphrase: string,
		options?: {
			includeSenderSignature?: boolean;
			multisignatureKeys?: {
				mandatoryKeys: Buffer[];
				optionalKeys: Buffer[];
			};
		},
	): Promise<Record<string, unknown>> {
		const txInput = input;
		const networkIdentifier = Buffer.from(this._nodeInfo.networkIdentifier, 'hex');
		const { publicKey, address } = getAddressAndPublicKeyFromPassphrase(passphrase);
		const accountHex = await this._channel.invoke<string>('app:getAccount', {
			address: address.toString('hex'),
		});
		const account = decodeAccount(Buffer.from(accountHex, 'hex'), this._schema);
		if (!txInput.moduleID) {
			if (!txInput.moduleName) {
				throw new Error('Missing moduleID and moduleName');
			}
			const registeredModule = this._nodeInfo.registeredModules.find(
				module => module.name === input.moduleName,
			);
			if (!registeredModule) {
				throw new Error(`Module corresponding to name ${txInput.moduleName} not registered.`);
			}
			txInput.moduleID = registeredModule.id;
		}
		if (typeof txInput.assetID !== 'number') {
			if (!txInput.assetName) {
				throw new Error('Missing assetID and assetName');
			}
			const registeredModule = this._nodeInfo.registeredModules.find(
				m => m.id === txInput.moduleID,
			);
			if (!registeredModule) {
				throw new Error(`Module corresponding to id ${txInput.moduleID} not registered.`);
			}
			const registeredAsset = registeredModule.transactionAssets.find(
				asset => asset.name === txInput.assetName,
			);
			if (!registeredAsset) {
				throw new Error(`Asset corresponding to name ${txInput.assetName} not registered.`);
			}
			txInput.assetID = registeredAsset.id;
		}
		if (typeof txInput.nonce !== 'bigint') {
			if (
				typeof account.sequence !== 'object' ||
				typeof (account.sequence as Record<string, unknown>).nonce !== 'bigint'
			) {
				throw new Error('Unsupported account type');
			}
			txInput.nonce = (account.sequence as { nonce: bigint }).nonce;
		}
		if (txInput.nonce < BigInt(0)) {
			throw new Error('Nonce must be greater or equal to zero');
		}
		if (!txInput.senderPublicKey) {
			txInput.senderPublicKey = publicKey;
		}
		// If signature is not set, assign empty array
		txInput.signatures = txInput.signatures ?? [];
		const assetSchema = getTransactionAssetSchema(txInput, this._schema);
		if (account.keys && (account.keys as MultiSignatureKeys).numberOfSignatures > 0) {
			return signMultiSignatureTransaction(
				assetSchema,
				txInput,
				networkIdentifier,
				passphrase,
				account.keys as MultiSignatureKeys,
				options?.includeSenderSignature,
			);
		}
		if (options?.multisignatureKeys && options?.includeSenderSignature) {
			return signMultiSignatureTransaction(
				assetSchema,
				txInput,
				networkIdentifier,
				passphrase,
				options.multisignatureKeys,
				options.includeSenderSignature,
			);
		}
		return signTransaction(assetSchema, txInput, networkIdentifier, passphrase);
	}

	public async get(id: Buffer): Promise<Record<string, unknown>> {
		const transactionHex = await this._channel.invoke<string>('app:getTransactionByID', {
			id: id.toString('hex'),
		});
		return decodeTransaction(Buffer.from(transactionHex, 'hex'), this._schema);
	}

	public async getFromPool(): Promise<Record<string, unknown>[]> {
		const transactionsHex = await this._channel.invoke<string[]>('app:getTransactionsFromPool');
		const decodedTransactions: Record<string, unknown>[] = [];
		for (const transactionHex of transactionsHex) {
			decodedTransactions.push(decodeTransaction(Buffer.from(transactionHex, 'hex'), this._schema));
		}
		return decodedTransactions;
	}

	public async sign(
		transaction: Record<string, unknown>,
		passphrases: string[],
		options?: {
			includeSenderSignature?: boolean;
			multisignatureKeys?: {
				mandatoryKeys: Buffer[];
				optionalKeys: Buffer[];
			};
		},
	): Promise<Record<string, unknown>> {
		const assetSchema = getTransactionAssetSchema(transaction, this._schema);
		const networkIdentifier = Buffer.from(this._nodeInfo.networkIdentifier, 'hex');
		const address = getAddressFromPublicKey(transaction.senderPublicKey as Buffer);
		const accountHex = await this._channel.invoke<string>('app:getAccount', {
			address: address.toString('hex'),
		});
		const account = decodeAccount(Buffer.from(accountHex, 'hex'), this._schema);
		if (account.keys && (account.keys as MultiSignatureKeys).numberOfSignatures > 0) {
			for (const passphrase of passphrases) {
				signMultiSignatureTransaction(
					assetSchema,
					transaction,
					networkIdentifier,
					passphrase,
					account.keys as MultiSignatureKeys,
					options?.includeSenderSignature,
				);
			}
			return transaction;
		}
		if (options?.multisignatureKeys && options?.includeSenderSignature) {
			for (const passphrase of passphrases) {
				signMultiSignatureTransaction(
					assetSchema,
					transaction,
					networkIdentifier,
					passphrase,
					options.multisignatureKeys as MultiSignatureKeys,
					options.includeSenderSignature,
				);
			}
			return transaction;
		}
		return signTransaction(assetSchema, transaction, networkIdentifier, passphrases[0]);
	}

	public async send(
		transaction: Record<string, unknown>,
	): Promise<{
		transactionId: string;
	}> {
		const encodedTx = encodeTransaction(transaction, this._schema);
		return this._channel.invoke<{
			transactionId: string;
		}>('app:postTransaction', { transaction: encodedTx.toString('hex') });
	}

	public decode<T = Record<string, unknown>>(transaction: Buffer): T {
		return decodeTransaction(transaction, this._schema) as T;
	}

	public encode(transaction: Record<string, unknown>): Buffer {
		return encodeTransaction(transaction, this._schema);
	}

	public computeMinFee(transaction: Record<string, unknown>): bigint {
		const assetSchema = getTransactionAssetSchema(transaction, this._schema);
		return computeMinFee(assetSchema, transaction);
	}

	public toJSON(transaction: Record<string, unknown>): Record<string, unknown> {
		const { asset: txAsset, ...txRoot } = transaction;
		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpId = txRoot.id;
		delete txRoot.id;

		const schemaAsset = getTransactionAssetSchema(txRoot, this._schema);
		const jsonTxAsset = codec.toJSON(schemaAsset, txAsset as Record<string, unknown>);
		const jsonTxRoot = codec.toJSON(this._schema.transaction, txRoot);

		const jsonTx = {
			...jsonTxRoot,
			asset: jsonTxAsset,
			id: Buffer.isBuffer(tmpId) ? tmpId.toString('hex') : tmpId,
		};

		return jsonTx;
	}

	public fromJSON(transaction: Record<string, unknown>): Record<string, unknown> {
		const { asset: txAsset, ...txRoot } = transaction;
		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpId = txRoot.id;
		delete txRoot.id;

		const schemaAsset = getTransactionAssetSchema(txRoot, this._schema);
		const txAssetObject = codec.fromJSON(schemaAsset, txAsset as Record<string, unknown>);
		const txRootObject = codec.fromJSON(this._schema.transaction, txRoot);

		const txObject = {
			...txRootObject,
			asset: txAssetObject,
			id: typeof tmpId === 'string' ? Buffer.from(tmpId, 'hex') : Buffer.alloc(0),
		};

		return txObject;
	}
}
