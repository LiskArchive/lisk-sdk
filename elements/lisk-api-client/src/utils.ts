/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { HashMap } from './api_types';

export const toQueryString = (obj: HashMap): string => {
	const parts = Object.keys(obj).reduce(
		(
			accumulator: ReadonlyArray<string>,
			key: string,
		): ReadonlyArray<string> => [
			...accumulator,
			`${encodeURIComponent(key)}=${encodeURIComponent(obj[key])}`,
		],
		[],
	);

	return parts.join('&');
};

const urlParamRegex = /{[^}]+}/;
export const solveURLParams = (url: string, params: HashMap = {}): string => {
	if (Object.keys(params).length === 0) {
		if (url.match(urlParamRegex) !== null) {
			throw new Error('URL is not completely solved');
		}

		return url;
	}
	const solvedURL = Object.keys(params).reduce(
		(accumulator: string, key: string): string =>
			accumulator.replace(`{${key}}`, params[key]),
		url,
	);

	if (solvedURL.match(urlParamRegex) !== null) {
		throw new Error('URL is not completely solved');
	}

	return encodeURI(solvedURL);
};
