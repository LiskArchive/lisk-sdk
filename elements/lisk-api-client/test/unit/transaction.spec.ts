/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { when } from 'jest-when';
import { getAddressAndPublicKeyFromPassphrase } from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import { Transaction } from '../../src/transaction';
import { nodeInfo, schema, tx, accountSchema } from '../utils/transaction';

describe('transaction', () => {
	let channelMock: any;
	let transaction: Transaction;
	const passphrases = [
		'trim elegant oven term access apple obtain error grain excite lawn neck',
		'faculty inspire crouch quit sorry vague hard ski scrap jaguar garment limb',
	];
	const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
	const { publicKey: publicKey1 } = getAddressAndPublicKeyFromPassphrase(passphrase1);
	const publicKey2 = Buffer.from(
		'fa406b6952d377f0278920e3eb8da919e4cf5c68b02eeba5d8b3334fdc0369b6',
		'hex',
	);
	const txHex =
		'0802100018362080ade2042a20dd4ff255fe04dd0159a468e9e9c8872c4f4466220f7e326377a0ceb9df2fa21a321d0880ade2041214654087c2df870402ab0b1996616fd3355d61f62c1a003a4079cb29dca7bb9fce73a1e8ca28264f779074d259c341b536bae9a54c0a2e4713580fcb192f9f15f43730650d69bb1f3dcfb4cb6da7d69ca990a763ed78569700';
	const accountHex =
		'0a14ab0041a7d3f7b2c290b5b834d46bdc7b7eb8581512050880c2d72f1a020800220208002a3b0a1a0a0a67656e657369735f3834180020850528003080a094a58d1d121d0a14ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151080a094a58d1d';
	const encodedTx = Buffer.from(txHex, 'hex');

	const validTransaction = {
		moduleID: 2,
		assetID: 0,
		nonce: BigInt('1'),
		fee: BigInt('10000000'),
		senderPublicKey: publicKey1,
		asset: {
			recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
			amount: BigInt('4008489300000000'),
			data: '',
		},
	};
	const txId = Buffer.from(tx.id, 'hex');

	beforeEach(() => {
		channelMock = {
			connect: jest.fn(),
			disconnect: jest.fn(),
			invoke: jest.fn(),
			subscribe: jest.fn(),
		};

		when(channelMock.invoke)
			.calledWith('app:getAccount')
			.mockResolvedValue(accountHex as never)
			.calledWith('app:getTransactionByID')
			.mockResolvedValue(txHex as never)
			.calledWith('app:getTransactionsFromPool')
			.mockResolvedValue([txHex] as never)
			.calledWith('app:postTransaction')
			.mockResolvedValue(txHex as never);

		transaction = new Transaction(channelMock, schema, nodeInfo);
	});

	describe('Transaction', () => {
		describe('constructor', () => {
			it('should initialize with channel', () => {
				expect(transaction['_channel']).toBe(channelMock);
			});
		});

		describe('get', () => {
			it('should invoke app:getTransactionByID', async () => {
				await transaction.get(txId);
				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:getTransactionByID', {
					id: txId.toString('hex'),
				});
			});
		});

		describe('getFromPool', () => {
			it('should invoke app:getTransactionsFromPool', async () => {
				await transaction.getFromPool();
				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:getTransactionsFromPool');
			});
		});

		describe('create', () => {
			describe('when called with a valid transaction', () => {
				it('should return created tx', async () => {
					const returnedTx = await transaction.create(validTransaction, passphrase1);
					expect(returnedTx.signatures).toHaveLength(1);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});

			describe('when called without module id and module name in input', () => {
				it('should throw error', async () => {
					await expect(
						transaction.create({ ...validTransaction, moduleID: undefined }, passphrase1),
					).rejects.toThrow('Missing moduleID and moduleName');
				});
			});

			describe('when called without asset id and asset name in input', () => {
				it('should throw error', async () => {
					await expect(
						transaction.create({ ...validTransaction, assetID: undefined }, passphrase1),
					).rejects.toThrow('Missing assetID and assetName');
				});
			});

			describe('when called with module name which does not exist', () => {
				it('should throw error', async () => {
					await expect(
						transaction.create(
							{ ...validTransaction, moduleID: undefined, moduleName: 'newModule' },
							passphrase1,
						),
					).rejects.toThrow('Module corresponding to name newModule not registered.');
				});
			});

			describe('when called with asset name which does not exist', () => {
				it('should throw error', async () => {
					await expect(
						transaction.create(
							{ ...validTransaction, assetID: undefined, assetName: 'newAsset' },
							passphrase1,
						),
					).rejects.toThrow('Asset corresponding to name newAsset not registered.');
				});
			});

			describe('when called without nonce in input and account does not support nonce either', () => {
				beforeEach(() => {
					(codec as any)['_compileSchemas'] = [];
				});
				afterEach(() => {
					(codec as any)['_compileSchemas'] = [];
				});
				it('should throw error', async () => {
					const updatedSchema = {
						...schema,
						account: {
							...schema.account,
							properties: {
								address: schema.account.properties.address,
								keys: schema.account.properties.keys,
							},
						},
					};
					transaction = new Transaction(channelMock, updatedSchema, nodeInfo);
					await expect(
						transaction.create({ ...validTransaction, nonce: undefined }, passphrase1),
					).rejects.toThrow('Unsupported account type');
				});
			});

			describe('when called with negative nonce in input', () => {
				it('should throw error', async () => {
					await expect(
						transaction.create({ ...validTransaction, nonce: BigInt(-2452) }, passphrase1),
					).rejects.toThrow('Nonce must be greater or equal to zero');
				});
			});

			describe('when called with nonce equal to zero in input', () => {
				it('should return created tx', async () => {
					const returnedTx = await transaction.create(
						{ ...validTransaction, nonce: BigInt(0) },
						passphrase1,
					);
					expect(returnedTx.signatures).toHaveLength(1);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});

			describe('when called without sender public key in input', () => {
				it('should return created tx', async () => {
					const returnedTx = await transaction.create(
						{ ...validTransaction, senderPublicKey: undefined },
						passphrase1,
					);
					expect(returnedTx.signatures).toHaveLength(1);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});

			describe('when called with multi-signature account in input', () => {
				it('should return created tx', async () => {
					const multisigAccount = {
						address: Buffer.from('ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815', 'hex'),
						token: { balance: BigInt('100000000') },
						sequence: { nonce: BigInt('0') },
						keys: {
							numberOfSignatures: 1,
							mandatoryKeys: [publicKey1],
							optionalKeys: [publicKey2],
						},
					};
					const multisigAccountHex = codec.encode(accountSchema, multisigAccount);
					when(channelMock.invoke)
						.calledWith('app:getAccount')
						.mockResolvedValue(multisigAccountHex.toString('hex') as never);
					const returnedTx = await transaction.create(validTransaction, passphrase1);
					expect(returnedTx.signatures).toHaveLength(2);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});

			describe('when called with optional keys in input', () => {
				it('should return created tx', async () => {
					const options = {
						includeSenderSignature: true,
						multisignatureKeys: {
							mandatoryKeys: [],
							optionalKeys: [publicKey2],
						},
					};
					const returnedTx = await transaction.create(validTransaction, passphrase1, options);
					expect(returnedTx.signatures).toHaveLength(2);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});
		});

		describe('sign', () => {
			describe('when called with a valid transation', () => {
				it('should return some signed transaction', () => {
					const returnedTx = transaction.sign(validTransaction, passphrases);
					expect(returnedTx).toBeDefined();
				});
			});

			describe('when called with multi-signature account in input', () => {
				it('should return created tx', async () => {
					const multisigAccount = {
						address: Buffer.from('ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815', 'hex'),
						token: { balance: BigInt('100000000') },
						sequence: { nonce: BigInt('0') },
						keys: {
							numberOfSignatures: 1,
							mandatoryKeys: [publicKey1],
							optionalKeys: [publicKey2],
						},
					};
					const multisigAccountHex = codec.encode(accountSchema, multisigAccount);
					when(channelMock.invoke)
						.calledWith('app:getAccount')
						.mockResolvedValue(multisigAccountHex.toString('hex') as never);
					const returnedTx = await transaction.sign(validTransaction, passphrases);
					expect(returnedTx.signatures).toHaveLength(2);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});

			describe('when called with optional keys in input', () => {
				it('should return created tx', async () => {
					const options = {
						includeSenderSignature: true,
						multisignatureKeys: {
							mandatoryKeys: [],
							optionalKeys: [publicKey2],
						},
					};
					const returnedTx = await transaction.sign(validTransaction, passphrases, options);
					expect(returnedTx.signatures).toHaveLength(2);
					expect(returnedTx.signatures).toMatchSnapshot();
				});
			});
		});

		describe('send', () => {
			it('should invoke app:postTransaction', async () => {
				const trxId = await transaction.send(tx);

				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:postTransaction', {
					transaction: txHex,
				});
				expect(trxId).toEqual(txHex);
			});
		});

		describe('decode', () => {
			it('should return decoded transaction', () => {
				const decodedTx = transaction.decode(encodedTx);
				expect(decodedTx).toMatchSnapshot();
			});
		});

		describe('encode', () => {
			it('should return encoded transaction', () => {
				const returnedTx = transaction.encode(tx);
				expect(returnedTx).toEqual(encodedTx);
			});
		});

		describe('computeMinFee', () => {
			it('should return some value', () => {
				const fee = transaction.computeMinFee(tx);
				expect(fee).toBeDefined();
			});
		});

		describe('toJSON', () => {
			it('should return decoded transaction in JSON', () => {
				const txAsJSON = transaction.toJSON(tx);
				expect(() => JSON.parse(JSON.stringify(txAsJSON))).not.toThrow();
			});
		});

		describe('fromJSON', () => {
			it('should return decoded transaction in JSON', () => {
				const txCopy = { ...tx };
				(txCopy as any).id = txId;
				const txAsJSON = transaction.toJSON(txCopy);
				const txAsObject = transaction.fromJSON(txAsJSON);

				expect(txAsObject).toEqual(txCopy);
			});
		});
	});
});
