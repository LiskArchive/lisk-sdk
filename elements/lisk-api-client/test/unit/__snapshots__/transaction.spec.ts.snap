// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction Transaction create when called with a valid transaction should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
]
`;

exports[`transaction Transaction create when called with multi-signature account in input should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
  "",
]
`;

exports[`transaction Transaction create when called with nonce equal to zero in input should return created tx 1`] = `
Array [
  "03430d5cedeab0db30c2b32e60920d74da02e0aafde8f786ac7153deb9cf78ba6ffc955f911ff66ce690c47206a23fa53c2e1ee3e1f4850ff8c37a8fe1a6d50c",
]
`;

exports[`transaction Transaction create when called with optional keys in input should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
]
`;

exports[`transaction Transaction create when called without sender public key in input should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
]
`;

exports[`transaction Transaction decode transaction from input as buffer should return decoded transaction 1`] = `
Object {
  "command": "transfer",
  "fee": 10000000n,
  "id": Object {
    "data": Array [
      116,
      198,
      128,
      147,
      238,
      181,
      55,
      166,
      114,
      181,
      57,
      160,
      183,
      70,
      212,
      192,
      169,
      102,
      248,
      94,
      221,
      18,
      148,
      116,
      213,
      248,
      112,
      41,
      174,
      250,
      16,
      185,
    ],
    "type": "Buffer",
  },
  "module": "token",
  "nonce": 54n,
  "params": Object {
    "amount": 10000000n,
    "data": "",
    "recipientAddress": Object {
      "data": Array [
        101,
        64,
        135,
        194,
        223,
        135,
        4,
        2,
        171,
        11,
        25,
        150,
        97,
        111,
        211,
        53,
        93,
        97,
        246,
        44,
      ],
      "type": "Buffer",
    },
  },
  "senderPublicKey": Object {
    "data": Array [
      221,
      79,
      242,
      85,
      254,
      4,
      221,
      1,
      89,
      164,
      104,
      233,
      233,
      200,
      135,
      44,
      79,
      68,
      102,
      34,
      15,
      126,
      50,
      99,
      119,
      160,
      206,
      185,
      223,
      47,
      162,
      26,
    ],
    "type": "Buffer",
  },
  "signatures": Array [
    Object {
      "data": Array [
        121,
        203,
        41,
        220,
        167,
        187,
        159,
        206,
        115,
        161,
        232,
        202,
        40,
        38,
        79,
        119,
        144,
        116,
        210,
        89,
        195,
        65,
        181,
        54,
        186,
        233,
        165,
        76,
        10,
        46,
        71,
        19,
        88,
        15,
        203,
        25,
        47,
        159,
        21,
        244,
        55,
        48,
        101,
        13,
        105,
        187,
        31,
        61,
        207,
        180,
        203,
        109,
        167,
        214,
        156,
        169,
        144,
        167,
        99,
        237,
        120,
        86,
        151,
        0,
      ],
      "type": "Buffer",
    },
  ],
}
`;

exports[`transaction Transaction decode transaction from input as hex should return decoded transaction 1`] = `
Object {
  "command": "transfer",
  "fee": 10000000n,
  "id": Object {
    "data": Array [
      116,
      198,
      128,
      147,
      238,
      181,
      55,
      166,
      114,
      181,
      57,
      160,
      183,
      70,
      212,
      192,
      169,
      102,
      248,
      94,
      221,
      18,
      148,
      116,
      213,
      248,
      112,
      41,
      174,
      250,
      16,
      185,
    ],
    "type": "Buffer",
  },
  "module": "token",
  "nonce": 54n,
  "params": Object {
    "amount": 10000000n,
    "data": "",
    "recipientAddress": Object {
      "data": Array [
        101,
        64,
        135,
        194,
        223,
        135,
        4,
        2,
        171,
        11,
        25,
        150,
        97,
        111,
        211,
        53,
        93,
        97,
        246,
        44,
      ],
      "type": "Buffer",
    },
  },
  "senderPublicKey": Object {
    "data": Array [
      221,
      79,
      242,
      85,
      254,
      4,
      221,
      1,
      89,
      164,
      104,
      233,
      233,
      200,
      135,
      44,
      79,
      68,
      102,
      34,
      15,
      126,
      50,
      99,
      119,
      160,
      206,
      185,
      223,
      47,
      162,
      26,
    ],
    "type": "Buffer",
  },
  "signatures": Array [
    Object {
      "data": Array [
        121,
        203,
        41,
        220,
        167,
        187,
        159,
        206,
        115,
        161,
        232,
        202,
        40,
        38,
        79,
        119,
        144,
        116,
        210,
        89,
        195,
        65,
        181,
        54,
        186,
        233,
        165,
        76,
        10,
        46,
        71,
        19,
        88,
        15,
        203,
        25,
        47,
        159,
        21,
        244,
        55,
        48,
        101,
        13,
        105,
        187,
        31,
        61,
        207,
        180,
        203,
        109,
        167,
        214,
        156,
        169,
        144,
        167,
        99,
        237,
        120,
        86,
        151,
        0,
      ],
      "type": "Buffer",
    },
  ],
}
`;

exports[`transaction Transaction sign when called with multi-signature account in input should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
  "746711c65d963b1fe8c6e40d64100548fb1080a943c2b53ac8466570f96c4687478a9307b45ad0172da134aa1eb9c44e2f11b3f08e790bdfd9247f3f89526508",
]
`;

exports[`transaction Transaction sign when called with optional keys in input should return created tx 1`] = `
Array [
  "d27e0d4c1504e47d06e85c510946fda57c12279da70e2bb1f3b2bc3760127368c03bbf443e562543e9827beb433e97614dff59fbfca5c5776b2fc6f237b77009",
  "746711c65d963b1fe8c6e40d64100548fb1080a943c2b53ac8466570f96c4687478a9307b45ad0172da134aa1eb9c44e2f11b3f08e790bdfd9247f3f89526508",
]
`;
